### ISSUE TITLE
950PR上跑fp8算子coredump，报错Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed

### ISSUE THREAD

**Issue by chaosisnotopen** [triton-lang/triton-ascend#317]
### Describe the bug

算子代码如下：
# Copyright © 2026 Huawei Technologies Co., Ltd.
import math
import torch
import triton
import triton.language as tl

DEVICE = (
    "cuda"
    if torch.cuda.is_available()
    else "npu"
    if torch.npu.is_available()
    else "cpu"
)

@triton.jit
def _per_token_group_quant_fp8(
    x_ptr,
    x_q_ptr,
    x_s_ptr,
    group_size,
    num_groups,
    y_num_columns,
    y_row_stride,
    eps,
    fp8_min: tl.constexpr,
    fp8_max: tl.constexpr,
    use_ue8m0: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
    N: tl.constexpr,
):
    groups_per_row = y_num_columns // group_size
    local_pid = tl.program_id(0)
    M = tl.num_programs(0)

    # ====================== 循环处理多个 group ======================
    for iN in range(N):
        # 正确计算全局 group ID
        g_id = local_pid + iN * M

        # 计算属于第几行、第几个group
        row = g_id // groups_per_row
        row_g_id = g_id % groups_per_row

        # ====================== 【关键】每次循环都重新计算指针，不使用 += ======================
        x_g_ptr = x_ptr + (row * y_row_stride) + (row_g_id * group_size)
        x_q_g_ptr = x_q_ptr + g_id * group_size
        x_s_g_ptr = x_s_ptr + g_id

        # 加载
        cols = tl.arange(0, BLOCK_SIZE)
        mask = (cols < group_size) & (g_id < num_groups)
        x = tl.load(x_g_ptr + cols, mask=mask, other=0.0)

        # 量化计算
        _absmax = tl.maximum(tl.max(tl.abs(x)), eps)
        scale_raw = _absmax * (1.0 / fp8_max)
        x_s = tl.math.exp2(tl.ceil(tl.log2(scale_raw))) if use_ue8m0 else scale_raw
        x_q = tl.clamp(x / x_s, fp8_min, fp8_max).to(x_q_ptr.dtype.element_ty)

        # 写回（每次都用新指针，不叠加）
        tl.store(x_q_g_ptr + cols, x_q, mask=mask)
        tl.store(x_s_g_ptr, x_s, mask=g_id < num_groups)

@triton.jit
def _per_token_group_quant_fp8_colmajor(
    x_ptr,
    x_q_ptr,
    x_s_ptr,
    group_size,
    num_groups,
    y_num_columns,
    y_row_stride,
    y_s_col_stride,
    eps,
    fp8_min: tl.constexpr,
    fp8_max: tl.constexpr,
    use_ue8m0: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
    N: tl.constexpr,
):
    groups_per_row = y_num_columns // group_size
    local_pid = tl.program_id(0)
    M = tl.num_programs(0)

    for iN in range(N):
        g_id = local_pid + iN * M
        row = g_id // groups_per_row
        row_g_id = g_id % groups_per_row

        # ====================== 【关键】每次循环重新计算，不 += ======================
        x_g_ptr = x_ptr + row * y_row_stride + row_g_id * group_size
        x_q_g_ptr = x_q_ptr + g_id * group_size

        blocks_per_row = groups_per_row
        scale_col = g_id % blocks_per_row
        scale_row = g_id // blocks_per_row

        # scale 指针正确计算
        scale_offset = scale_col * y_s_col_stride + scale_row
        x_s_g_ptr = x_s_ptr + scale_offset

        cols = tl.arange(0, BLOCK_SIZE)
        mask = (cols < group_size) & (g_id < num_groups)

        x = tl.load(x_g_ptr + cols, mask=mask, other=0.0)
        _absmax = tl.maximum(tl.max(tl.abs(x)), eps)
        scale_raw = _absmax * (1.0 / fp8_max)
        x_s = tl.math.exp2(tl.ceil(tl.log2(scale_raw))) if use_ue8m0 else scale_raw
        x_q = tl.clamp(x / x_s, fp8_min, fp8_max).to(x_q_ptr.dtype.element_ty)

        tl.store(x_q_g_ptr + cols, x_q, mask=mask)
        tl.store(x_s_g_ptr, x_s, mask=g_id < num_groups)

def triton_per_token_group_quant_fp8(
    x: torch.Tensor,
    group_size: int,
    dtype: torch.dtype,
    column_major_scales: bool = False,
    block_size=None,
    eps: float = 1e-10,
    use_ue8m0: bool = False,
) -> tuple[torch.Tensor, torch.Tensor]:
    assert x.shape[-1] % group_size == 0, (
        f"the last dimension of `x` {x.shape[-1]} must be divisible "
        f"by `group_size` {group_size}"
    )
    assert x.stride(-1) == 1, "`x` must be contiguous on the last dimension"

    x = x.to(DEVICE)
    fp8_min = torch.finfo(dtype).min
    fp8_max = torch.finfo(dtype).max

    # 输出tensor分配
    x_q = torch.empty(x.shape, device=x.device, dtype=dtype)
    if column_major_scales:
        shape = x.shape[:-2] + (x.shape[-1] // group_size, x.shape[-2])
        x_s = torch.empty(shape, device=x.device, dtype=x.dtype).permute(-1, -2)
    else:
        shape = x.shape[:-1] + (x.shape[-1] // group_size,)
        x_s = torch.empty(shape, device=x.device, dtype=x.dtype)

    num_groups = x.numel() // group_size

    # Grid 配置（NPU自动获取VectorCore数量）
    if DEVICE == "cuda":
        block_size = triton.next_power_of_2(group_size)
        N = 1
        grid = (num_groups,)
    else:
        block_size = group_size
        import triton.runtime.driver as driver

        device_idx = torch.npu.current_device()
        core_num = driver.active.utils.get_device_properties(device_idx)[
            "num_vectorcore"
        ]
        grid = (core_num,)
        N = (num_groups + core_num - 1) // core_num

    # 启动kernel
    if column_major_scales:
        _per_token_group_quant_fp8_colmajor[grid](
            x,
            x_q,
            x_s,
            group_size,
            num_groups,
            x.shape[1],
            x.stride(0),
            x_s.stride(1),
            eps,
            fp8_min,
            fp8_max,
            use_ue8m0,
            block_size,
            N,
        )
    else:
        # 【修复】统一使用grid启动，不写死(M,)
        _per_token_group_quant_fp8[grid](
            x,
            x_q,
            x_s,
            group_size,
            num_groups,
            x.shape[1],
            x.stride(0),
            eps,
            fp8_min,
            fp8_max,
            use_ue8m0,
            block_size,
            N,
        )

    return x_q, x_s

测试代码如下：

NUM_TOKENS = [7, 2050]
DIMS = [512, 4096, 5120, 13824]
GROUP_SIZES = [64, 128, 512]
# DTYPES = [torch.float16, torch.float8_e5m2, torch.float8_e4m3fn]
DTYPES = [torch.float8_e5m2, torch.float8_e4m3fn]
COLUMN_MAJOR_SCALES = [True, False]
SEEDS = [123]

# 精度标准
RTOL = {torch.float32: 1e-3, torch.float16: 1e-1}
ATOL = {torch.float32: 1e-4, torch.float16: 1e-1}

# 性能测试标准
PERFORMANCE_THRESHOLD = 0.9  # 0.9 * AscendC
WARMUP = 0
ITER = 1

def benchmark_perf(num_tokens, ndim, group_size, block_size, dtype, provider):
    print("====DTYPE={}".format(dtype))
    shape = (num_tokens, ndim)
    x = torch.rand(shape, dtype=torch.float32).to(dtype).to(DEVICE)

    USE_AUTOTUNE = block_size == 0

    if provider == "torch":
        ms = triton.testing.do_bench(
            lambda: torch_per_token_group_quant_fp8(x, group_size, dtype)
        )
    if provider == "triton":
        if USE_AUTOTUNE:
            ms = triton.testing.do_bench(
                lambda: triton_per_token_group_quant_fp8(x, group_size, dtype)
            )
        else:
            ms = triton.testing.do_bench(
                lambda: triton_per_token_group_quant_fp8(
                    x, group_size, dtype, block_size=block_size
                )
            )

    return ms

configs = []
for _num, _dim in [[2050, 13824], [500, 5120], [7, 512]]:
    for _group_size in GROUP_SIZES:
        for _dtype in DTYPES:
            configs.append(
                triton.testing.Benchmark(
                    # argument names to use as an x-axis for the plot
                    x_names=[
                        "block_size",
                    ],
                    x_vals=[
                        32,
                        64,
                        96,
                        128,
                        256,
                        512,
                    ],  # different possible values for `x_name`
                    line_arg="provider",
                    # argument name whose value corresponds to a different line in the plot
                    # possible values for `line_arg``
                    line_vals=["torch", "triton"],
                    # label name for the lines
                    line_names=["torch", "triton"],
                    # line styles
                    styles=[("green", "-"), ("blue", "-")],
                    ylabel="runtime(ms)",  # label name for the y-axis
                    plot_name=f"per_token_group_quant_fp8-{_num}-{_dim}-{_group_size}-{_dtype}",
                    # name for the plot. Used also as a file name for saving the plot.
                    args={
                        "num_tokens": _num,
                        "ndim": _dim,
                        "dtype": _dtype,
                        "group_size": _group_size,
                    },
                )
            )

@triton.testing.perf_report(configs)
def do_bench_perf(num_tokens, ndim, group_size, block_size, dtype, provider):
    return benchmark_perf(num_tokens, ndim, group_size, block_size, dtype, provider)

if __name__ == "__main__":
    # 不输入参数, do_bench跑指定切块的性能
    if len(sys.argv) == 1:
        do_bench_perf.run(show_plots=True, print_data=True, save_path=".")

### Environment details

triton3.2.1
950PR
